File: js/basicType.js
const common = require('./schemaCommon');

function anyType(value, path){
    return value;
}

function stringType(value, path){
    path = path || [];
    if (value === null) {
        return null;
    }
    if (!common.isString(value)){
        throw new Error(common.getMessage(path, 'Should be a string'));
    }
    return value;
}

function integerType(value, path){
    path = path || [];
    if (value === null) {
        return null;
    }
    if (!common.isInteger(value)){
        throw new Error(common.getMessage(path, 'Should be an integer'));
    }
    return value;
}

function booleanType(value, path){
    path = path || [];
    if (value === null) {
        return null;
    }
    if (!common.isBoolean(value)){
        throw new Error(common.getMessage(path, 'Should be a boolean'));
    }
    return value;
}

function validateStringMap(value, path){
    if (!common.isDict(value)){
        throw new Error(common.getMessage(path, 'Should be an object'));
    }
    for (let prop in value){
        if (value.hasOwnProperty(prop)){
            if (!common.isString(value[prop])){
                let ext_path = path.concat([prop]);
                throw new Error(common.getMessage(ext_path, 'Should be a string'));
            }
        }
    }
}

function stringMap(value, path){
    path = path || [];
    if (value === null) {
        return null;
    }
    validateStringMap(value, path);
    return value;
}

module.exports = {
    anyType: anyType,
    stringType: stringType,
    integerType: integerType,
    booleanType: booleanType,
    stringMap: stringMap
};

File: python3/src/basic_type.py
from .schema_dsl_common import *


def any_type(value, path):
    return value


def string_type(value, path):
    path = path or []
    if value is None:
        return None
    if not isinstance(value, str):
        raise TypeError(get_message(path, 'Should be a string'))
    return value


def integer_type(value, path):
    path = path or []
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, int):
        raise TypeError(get_message(path, 'Should be an integer'))
    return value


def number_type(value, path):
    path = path or []
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise TypeError(get_message(path, 'Should be a number'))
    return value


def boolean_type(value, path=None):
    path = path or []
    if value is None:
        return None
    if not isinstance(value, bool):
        raise TypeError(get_message(path, 'Should be a boolean'))
    return value


def _validate_string_map(value, path):
    if not isinstance(value, dict):
        raise TypeError(get_message(path, 'Should be an object'))
    for k, v in value.items():
        if not isinstance(v, str):
            raise TypeError(get_message(path + [k], 'Should be a string'))


def string_map(value, path):
    path = path or []
    if value is None:
        return None
    _validate_string_map(value, path)
    return value
